### PR TITLE
HHVM Extension Support

### PR DESCRIPTION
**Pull request by RyanGordon**
This adds support for running gRPC with HHVM as a native extension. This ports over all of the methods and code from the PHP extension with the same interface. This is apart of getting Cloud Spanner to run on HHVM using the official PHP library which requires gRPC.

This still is a WIP, and not ready for production.

### CODE REVIEW

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by RyanGordon** [comment]
grpc-testing really wants to test this :)

Past that, thanks everyone for taking a look at this. I have a few questions on this PR:

1) Should I setup the directory structure like this? Since a lot of the code can be shared with the php directory (like all the tests) it may be better to place this under the PHP folder in some way. I also haven't added any different tests here since all the same ones from the php library can be used.

2) I have been working through a few remaining segfaults and memory issues and have gotten stuck. For instance the following stack trace (https://gist.github.com/RyanGordon/d812dbe22d599ce5d54365c5f64b205c) has plagued me and may be related to calling `vm_call_user_func` in some form and memory not being set properly. There's also a separate segfault that happens and I haven't been able to locate why yet: https://gist.github.com/RyanGordon/31246ced243d72bf2fbd1fa48f4cb3b5

Any theories or help on the segfault/error would be greatly appreciated.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by RyanGordon** [comment]
I was able to fix the segfault and memory corruption issues mentioned above; I'm still working through a few more issues but those should hopefully work soon

**Comment by RyanGordon** [comment]
A quick update: I've fixed any obvious issues and now have connection, and querying and related things working with this. Next, I'm going to be going through and evaluating for thread safety and testing it, then I'll ensure all unit tests pass and lastly I'll do a round of memory leak testing.

I'm still curious for feedback on how this should be structured? Should I merge it in with the PHP directory in some form or copy tests over, etc?

**Comment by stanley-cheung** [comment]
@RyanGordon Thanks for the contribution! This is amazing. Great work! We are still taking a look at this and are discussing folder structure. We will have an update soon. Thanks!

**Comment by RyanGordon** [comment]
Just wanted to give an update on this: The extension is now stable other than one or more memory leaks I am tracking down. I also have a few changes to make to get all of the unit tests to pass but they're mostly around the type of exception that the unit test expects from the extension.

**Comment by billf** [comment]
> Please have them confirm that here in the pull request.

i'm ok with that.

**Comment by stanley-cheung** [comment]
Hi @RyanGordon, thanks again for your contribution! I have a couple of questions

1. I noticed that you removed the `lib/Grpc` code (which was in Hack?). Does that mean the HHVM extension will work with the PHP `lib/Grpc` code? And I suppose same for all the unit tests as well? So this PR really only needs the HHVM C extension?

2. What's the use of the `composer.json` in this context? That is only used to pull in other packages dependency to run the PHP lib code right? Coz there's no PHP code, per se, to submit this as a composer package right?

3. We are still deciding on the folder structure. But now that it appears that this PR will only have the HHVM C extension as "new code". I am leaning towards putting this in the `grpc/src/php/ext` subtree. What do you think?

**Comment by RyanGordon** [comment]
Hi @stanley-cheung 

1. Yes, it turns out the code in `lib/Grpc` doesn't need any changes for it to run in hack and the header isn't necessary unless we're adding typing information. I removed the folder for those reasons. I do have a separate file to run HHVM unit tests locally, since I am not sure how to setup the hhvm framework for the CI. It also appears the `autoload.php` that the php unit tests reference isn't in this repo, so I just used my own autoloader instead to run tests.

2. It does look like `composer.json` is indeed not needed in this PR (No differences from the PHP composer package); I will remove it.

3. That sounds correct to me. Should we put it under `grpc/src/php/ext/hhvm`?

**Comment by RyanGordon** [comment]
Hello @stanley-cheung,

I've been able to get the extension reliably working up to a few threads, and then I hit a race condition:

```
(gdb) bt
#0  0x000000000130891e in void HPHP::MemoryManager::refreshStatsImpl<true>(HPHP::MemoryUsageStats&) ()
#1  0x00000000033400c0 in HPHP::MemoryManager::newSlab(unsigned int) ()
#2  0x000000000333fbda in HPHP::MemoryManager::mallocSmallSizeSlow(unsigned int, unsigned int) ()
#3  0x0000000000d5b810 in HPHP::SystemLib::AllocStdClassObject() ()
#4  0x00007fc9ca9d96fb in HPHP::plugin_get_metadata (ptr=0x7fc9978a96f0, context=..., cb=0x7fc9ca608670, user_data=0x7fc98466ba00) at hhvm/grpc/src/hhvm/ext/grpc/call_credentials.cpp:122
#5  0x00007fc9ca609cd9 in ?? () from /usr/local/lib/libgrpc.so.4
#6  0x00007fc9ca60c183 in ?? () from /usr/local/lib/libgrpc.so.4
#7  0x00007fc9ca609659 in ?? () from /usr/local/lib/libgrpc.so.4
#8  0x00007fc9ca614e77 in ?? () from /usr/local/lib/libgrpc.so.4
#9  0x00007fc9ca615698 in ?? () from /usr/local/lib/libgrpc.so.4
#10 0x00007fc9ca615e26 in ?? () from /usr/local/lib/libgrpc.so.4
#11 0x00007fc9ca5bbf46 in grpc_combiner_continue_exec_ctx () from /usr/local/lib/libgrpc.so.4
#12 0x00007fc9ca5c97a8 in grpc_exec_ctx_flush () from /usr/local/lib/libgrpc.so.4
#13 0x00007fc9ca5c9b1a in ?? () from /usr/local/lib/libgrpc.so.4
#14 0x00007fc9ca6304b7 in ?? () from /usr/local/lib/libgrpc.so.4
#15 0x00007fc9d98e3dc5 in start_thread (arg=0x7fc982ffe700) at pthread_create.c:308
#16 0x00007fc9d417973d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

I've dug in a bunch and it appears that essentially HHVM is not thread safe within a request. Here gRPC is calling `plugin_get_metadata` (https://github.com/grpc/grpc/pull/11553/files#diff-b162b6e5f93a243a83db41b6aa4fa912R119) from a different thread, which is in turn calling HHVM application code which is not thread-safe within the context of a single request and so a segfault happens in a race. It appears as if this behavior will [not be supported](https://github.com/facebook/hhvm/issues/2827#issuecomment-172046327) by HHVM.

My question is, is there a way to workaround this behavior in gRPC? Ideally, I would like any gRPC calls to come back to the same thread that the HHVM request is running on, rather than executing any HHVM application code under a different thread. Is this possible?

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by stanley-cheung** [comment]
Hi @RyanGordon What you mentioned seems like the same issue as https://github.com/grpc/grpc/issues/7608.

**Comment by RyanGordon** [comment]
Ah yes, that does look very similar and your analysis agrees with mine per https://github.com/grpc/grpc/issues/7608#issuecomment-239326444

I don't suppose this mechanism has been implemented yet?

**Comment by stanley-cheung** [comment]
Yea the additional mechanism to make sure the `plugin_get_metadata` get executed back in the main thread hasn't been implemented yet.

**Comment by stanley-cheung** [comment]
At the time when we discover that issue, we can only reproduce that when we compile PHP 5.6 manually with the `enable-maintainer-zts` flag, which according to PHP's own documentation is not meant to be used for production. So we didn't think it was a high priority fix.

**Comment by RyanGordon** [comment]
I see; At the moment it seems this HHVM gRPC extension is unstable without this fix since HHVM is essentially "ZTS by default". Can we work together to implement this fix? I'd be happy to do the majority of the changes, but I'm very unfamiliar with the threading internals of the grpc core and don't really know where to start / what the fix would look like

**Comment by stanley-cheung** [comment]
How about something like this? In your HHVM extension, when you construct your `plugin_get_metadata` callback, don't have the callback immediately call into PHP-land to fetch the token. Instead, put all the information into a struct, and enqueue this into a separate queue you maintain. Now in the your main thread, the one that's waiting on `grpc_completion_queue_pluck`, occasionally go check on that new queue to see if there's an item there. (I believe you can set a timeout on `grpc_completion_queue_pluck`). If there is, do your actual `plugin_get_metadata` work there (i.e. go into PHP-land `user_call_func` to fetch the actual token). 

This new queue is a separate construct (not a GRPC concept per se) that you maintain in your HHVM extension.

**Comment by RyanGordon** [comment]
Hi @stanley-cheung,

That's a good idea; I've tried to implement something like it but unfortunately it's being more complicated than I would like. HHVM has a thread per request but everything within that thread must be serialized (i.e. no more than 1 thread per request). I think I have a way to handle this. Also it appears that sometimes the request will be run on the correct thread (the same one as the HHVM request thread) so we have to make sure we don't do the threading part in those situations

**Comment by RyanGordon** [comment]
Hi @stanley-cheung, I reimplemented my fix and it appears to be working stably now! I need to remove a couple of locks and fix a couple more memory leaks i introduced. Past that I just need to make the unit tests pass and then this should be good to start reviewing :+1:

**Comment by RyanGordon** [comment]
Hi @stanley-cheung 

I have officially run a stress test for 24 hours on this extension and there were no crashes for the duration of the test. This morning I fixed a final memory leak and memory appears to be stable and will continue to test for the next 24 hours.

While I am working on getting unit tests to pass where possible, would it be possible to get a CR on this in the next few days? I'm looking forward to getting this completed!

File: src/php/ext/grpc/hhvm/LICENSE [hunk 1]
```diff
@@ -0,0 +1,186 @@
+
```
Comment (ctiller) [inline src/php/ext/grpc/hhvm/LICENSE]: @jayantkolhe for explicit approval for the LICENSE

**Comment by ctiller** [comment]
I'm strongly opposed to the massive duplication of code that occurs in this change.
This is not maintainable, and a better solution will need to be found.

File: src/php/ext/grpc/hhvm/LICENSE (hunk 1 above)
Comment (RyanGordon) [inline src/php/ext/grpc/hhvm/LICENSE]: Should I just remove this file since there is already a LICENSE in the php directory?

**Comment by RyanGordon** [comment]
Hi @ctiller,

I'm happy to reduce duplication where possible. Unfortunately there are many changes that needed to be made as this is essentially for an entirely different language, as if we were writing the python extension for instance. Most of the code is in fact different for HHVM and not just duplication.

Are there specific places that you would like to see de-duplicated?

**Comment by ctiller** [comment]
Actually, can I pull that objection.

My initial read-through I thought I'd seen files from src/core/... copied into this tree -- which would be bad. I suspect they've been copied from the PHP tree (reasonable) and the internal extension function names just happen to alias some internal core function names (which again, is reasonable).

**Comment by RyanGordon** [comment]
Hi @stanley-cheung, et al; Would it be possible to get a CR on this relatively soon? We're working on the php unit tests and hoping to get those fixed up within the next week or so.

**Comment by stanley-cheung** [comment]
Sure, I will take a look early next week

**Comment by RyanGordon** [comment]
Hi @stanley-cheung, et al,

We're looking into a latency issue / edge case. We wrote it off before as something else but we're now investigating a latency issue in the extension. This is flow to see this odd behavior happen:

*Restart HHVM Service*
Request 1 (Thread A) against HHVM
- Run query #1 on Cloud Spanner
-- Request a new OAuth Token & cache it for future requests on Thread A (~50ms)
-- Create a new Cloud Spanner session & cache it for future requests on Thread A (~400ms)
-- Run actual query (~20-30ms)
- Run query #2 on Cloud Spanner
-- Reuse OAuth Token (<1ms)
-- Reuse Cloud Spanner Session (<1ms)
-- Run actual query (~8ms)

Request 2 (Thread A, same as before) against HHVM
- Run query #1 on Cloud Spanner
-- Get cached OAuth Token (~1ms)
-- Get cached Spanner session (<1ms)
- Run actual query (200ms) **(??? Where is this latency coming from?)**
- Run query #2 On Cloud Spanner
-- Reuse OAuth Token (<1ms)
-- Reuse Cloud Spanner Session (<1ms)
- Run actual query (8ms)

Query #1 is like this: `SELECT * FROM table WHERE column='ABC-DEF-0123'`
Query #2 is like this: `SELECT * FROM table WHERE column='ABC-DEF-0124'` (Just a different input)

I traced down the majority of the latency (67% of it) coming from something within: https://github.com/quizlet/grpc/blob/hhvm/src/php/ext/grpc/hhvm/call.cpp#L344 which is just waiting on stuff to happen in `grpc_completion_queue_pluck`

I'm curious if there is some documentation or explanation of what the inner workings of `grpc_completion_queue_pluck` or if anything might explain this behavior? I suspect it's just doing something like opening a new connection and doing some initialization step and that's causing the extra latency.

**Comment by stanley-cheung** [comment]
Hi @RyanGordon you can try enabling the C core tracing `GRPC_VERBOSITY=DEBUG` and `GRPC_TRACE=ALL` and see what's happening within a `pluck` call.

It could be because it's establishing the underlying `grpc_channel` connection. We are working on #11878 in parallel to save the underlying `grpc_channel` object so that we don't need to re-create the object for every request

**Comment by RyanGordon** [comment]
Ah very nice! I'll checkout both of those things and see what I can find

**Comment by RyanGordon** [comment]
Hi @stanley-cheung, I implemented the HHVM equivalent changes for persisting channels here: 11213749c26cd996f65410a75dfc1bc353aaa084 It looks like it brings down the latency from ~200ms to ~50ms! I'm going to dig in a little more to see what's going on with the extra 40ms but this is a great improvement

**Comment by RyanGordon** [comment]
Okay it looks like of the 50ms, about half is because we're using the regular PHP Protobuf library instead of the C optimized version and things like `\GPBMetadata\Google\Spanner\V1\Spanner::initOnce();` are being called at the beginning of every request. We're looking at getting the protobuf PHP C extension running on HHVM

**Comment by stanley-cheung** [comment]
Yes currently the PHP implementation of google/protobuf perform 10x-20x times worse than the C implementation. The gap might narrow in the future but the C implementation is a much safer bet.

**Comment by RyanGordon** [comment]
An update - We've went through and refactored the code for quality and made significant progress on stability. We however have been continuing to see a epoll1 hang manifesting itself after hours of stress testing. Here is a backtrace of the issue: https://gist.github.com/RyanGordon/f4117345ad468da5cad6319250821fc4

Specifically, threads 6 and 13 exhibit a hanging issue where they either hang on a gpr_cv_wait or do_epoll_wait. I've done a bunch of tracing in GDB and it appears that call.c:receiving_stream_ready_in_call_combiner isn't being scheduled or executed when it typically would.

@sreecha @markdroth or anyone else, are there any suggestions for what I should be looking more closely at through gdb or what might in general be going on?

**Comment by markdroth** [comment]
I haven't looked at this code at all, but I did notice that this code uses the plugin metadata API, so I wonder if it's suffering from the problem being fixed in #12374.

Try making sure that you're always calling the metadata plugin callback from a different call stack than the one that called out to you.

**Comment by RyanGordon** [comment]
@markdroth I saw that too and wondered if the same situation might be causing this. We _definitely_ call the callback from the same one called out to us often. Is there a preference between the async and sync versions? Is there a performance difference?

**Comment by markdroth** [comment]
We don't necessarily recommend one over the other.  The sync version may offer slightly better performance (although most users in high-level languages probably won't notice), but it is limited as to the number of metadata entries that can be returned (although the limit is high enough that most users won't care).  So it probably doesn't make much difference.

If you do call the callback from the same thread using the current API, you can indeed trigger a deadlock.  I can't say for sure whether that's actually happening, but the possibility is definitely there, because the exec_ctx handling is unsafe.  I suggest patching in my PR and updating your code to use the new API to see if that makes a difference.

**Comment by RyanGordon** [comment]
Great - We will go ahead and try that. For simplicity I think we'll try using the sync version instead

**Comment by fengli79** [comment]
@RyanGordon Thanks for keep improve this PR. Please let me know once it's ready for another round review.

**Comment by RyanGordon** [comment]
@fengli79 Thanks! I will do; I expect this will be ready for review in about a week. I may also close and re-open this as a new PR and squash the commits down so the PR is cleaner

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @RyanGordon isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf
* User @billf isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf
* User @bryanzim isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

**Pull request closed without merging**